lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +802,460 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+class GitRef(object):
+    """A wrapper that knows how to format git refspecs in multiple contexts."""
+    _known_types = ['branch', 'commit', 'tag']  # type: ClassVar[List[str]]
+    _ref_type = None                            # type: str
+    _ref = None                                 # type: str
+
+    @classmethod
+    def known_types(cls):
+        # type: () -> List[str]
+        return cls._known_types
+
+    @property
+    def ref_type(self):
+        # type: () -> str
+        return self._ref_type
+
+    def __init__(self, ref_type, ref):
+        # type: (str, str) -> None
+        if ref_type not in self.known_types():
+            # This is made a TypeError and not an InvalidGitRef because it cannot arise
+            # from invalid user input.
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types(), ref_type))
+        self._ref_type = ref_type
+        if not isinstance(ref, six.string_types):
```

Comment:
Should we do both checks first before setting attributes on either ref_type or ref?

Reply:
I find that logic tougher to follow in general, and more prone to getting subtly broken by someone else editing this code later, in comparison to a sequence of `[check->set, check->set]`. But I have actually in the meantime worked around the need to address this by making `ref_type` into a `ClassVar`!